### PR TITLE
Feature decommission

### PR DESCRIPTION
**Pull request by alinadonisa**
- remove unnecessary refs from scale package and add cockroachstatefullset.go file
- integrate decommission logic
- add  e2e test for decommission  
- add validation for min 3 nodes on CRD spec

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -3,6 +3,7 @@ module github.com/cockroachdb/cockroach-operator
 go 1.14
 
 require (
+	github.com/DataDog/datadog-go v4.2.0+incompatible // indirect
```
Comment (chrislovecnm) [inline go.mod:6]: @alinadonisa remove the datadog code please

File: pkg/scale/cockroach_statefulset.go [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+
+package scale
+
+import (
+	"context"
+	"time"
+
+	"github.com/cenkalti/backoff"
+	"github.com/cockroachdb/errors"
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer"
```
Comment (chrislovecnm) [inline pkg/scale/cockroach_statefulset.go:25]: Please remove the datadog code use.

**Review by chrislovecnm — COMMENTED** [review]
Some stuff I noticed

File: pkg/scale/cockroach_statefulset.go (hunk 2 above)
Comment (alinadonisa) [reply pkg/scale/cockroach_statefulset.go]: done :)

File: api/v1alpha1/cluster_types.go [hunk 3]
```diff
@@ -33,6 +33,7 @@ type CrdbClusterSpec struct {
 	// (Required) Number of nodes (pods) in the cluster
 	// +operator-sdk:csv:customresourcedefinitions:type=spec,displayName="Number of nodes",xDescriptors="urn:alm:descriptor:com.tectonic.ui:podCount"
 	// +required
+	// +kubebuilder:validation:Minimum=3
```
Comment (chrisseto) [inline api/v1alpha1/cluster_types.go:36]: hmmm, I think we still want to support making single node clusters?

We just can't support 2 nodes or go from 3 nodes to 1.

File: pkg/actor/decommission.go [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+	}
+	status := &ss.Status
+
+	if status.CurrentReplicas == 0 || status.CurrentReplicas < status.Replicas {
+		log.Info(" decommission statefulset does not have all replicas up")
+		return NotReadyErr{Err: errors.New("decommission statefulset does not have all replicas up")}
+	}
+
+	replicas := uint(cluster.Spec().Nodes)
+	if replicas < 3 {
```
Comment (chrisseto) [inline pkg/actor/decommission.go:82]: `<=`?

File: pkg/actor/decommission.go [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+		log.Info("We cannot decommission if there are less than 3 nodes", "nodes", replicas)
+		return errors.New("decommission with less than 3 nodes is not supported")
+	}
+	log.Info("replicas decommisioning", "status.CurrentReplicas", status.CurrentReplicas, "expected", cluster.Spec().Nodes)
+	if status.CurrentReplicas > cluster.Spec().Nodes {
```
Comment (chrisseto) [inline pkg/actor/decommission.go:87]: nit: if there's a short circuit case prefer to handle that with an early return for readability.

```go
if status.CurrentReplicas < cluster.Spec().Node {
    // nothing to do
    return nil
}
```

File: pkg/clustersql/settings.go [hunk 6]
```diff
@@ -0,0 +1,137 @@
+/*
```
Comment (chrisseto) [inline pkg/clustersql/settings.go:1]: This and the following file are already in the [`scale` package](https://github.com/cockroachdb/cockroach-operator/blob/master/pkg/scale/timeout.go)

File: pkg/scale/cockroach_statefulset.go [hunk 7]
```diff
@@ -0,0 +1,192 @@
...
+}
+
+// WaitUntilHealthy blocks until the target stateful set has exactly `scale` healthy replicas.
+func (c *CockroachStatefulSet) WaitUntilHealthy(ctx context.Context, scale uint) error {
+	return WaitUntilStatefulSetIsReadyToServe(ctx, c.ClientSet, c.Namespace, c.Name, int32(scale))
+}
+
+// WaitUntilStatefulSetIsRunning waits until the given statefulset has all pods scheduled and running but not necessarily healthy nor ready
+func WaitUntilStatefulSetIsRunning(ctx context.Context, clientset kubernetes.Interface, namespace string, name string) error {
+	// span, _ := tracer.StartSpanFromContext(ctx, "WaitUntilStatefulSetIsRunning")
```
Comment (chrisseto) [inline pkg/scale/cockroach_statefulset.go:103]: remove tracing?

File: pkg/scale/cockroach_statefulset.go [hunk 8]
```diff
@@ -0,0 +1,192 @@
...
+}
+
+//WaitUntilStatefulSetIsReadyToServe func
+func WaitUntilStatefulSetIsReadyToServe(
+	ctx context.Context,
+	clientset kubernetes.Interface,
+	namespace, name string,
+	numReplicas int32) error {
+
+	// span, _ := tracer.StartSpanFromContext(ctx, "WaitUntilStatefulSetIsReadyToServe")
```
Comment (chrisseto) [inline pkg/scale/cockroach_statefulset.go:162]: remove tracing?

File: pkg/scale/persistent_volume_pruner_test.go [hunk 9]
```diff
@@ -200,7 +201,8 @@ func TestPersistentVolumePruner_Prune(t *testing.T) {
 	for _, tc := range testCases {
 		t.Run(tc.Name, func(t *testing.T) {
 			ctx := context.Background()
-			logger := zaptest.NewLogger(t)
+			//logger := zaptest.NewLogger(t)
```
Comment (chrisseto) [inline pkg/scale/persistent_volume_pruner_test.go:204]: nit: remove commented code?

**Review by chrisseto — APPROVED** [review]
LGTM minus some nits and the inability to create single node clusters.

Creating small test clusters is a case that we should continue to support.

**Review by chrisseto — APPROVED** [review]
LGTM

**Pull request merged**